src/lib/integration_test_lib/error_monad.ml

```diff
@@ -0,0 +1,167 @@
+open Core_kernel
+open Async_kernel
+
+module Hard_fail = struct
+  type t =
+    { hard_error: Error.t
+          [@equal
+            fun a b ->
+              String.equal (Error.to_string_hum a) (Error.to_string_hum b)]
+    ; soft_errors: Error.t list
+          [@equal
+            List.equal (fun a b ->
+                String.equal (Error.to_string_hum a) (Error.to_string_hum b) )]
+    }
+  [@@deriving eq, sexp_of, compare]
+end
+
+module Accumulator = struct
+  type 'a t =
+    { computation_result: 'a
+    ; soft_errors: Error.t list
+          [@equal
+            List.equal (fun a b ->
+                String.equal (Error.to_string_hum a) (Error.to_string_hum b) )]
+    }
+  [@@deriving eq, sexp_of, compare]
+end
+
+module ErrorReporting = Monad.Make (struct
+  type 'a t = ('a Accumulator.t, Hard_fail.t) Deferred.Result.t
+
+  let return a =
+    Deferred.return (Ok {Accumulator.computation_result= a; soft_errors= []})
```

Comment:
A convenient shorthand for this `Deferred.return (Ok x)` is `Deferred.Result.return x`. If you had a function `Accumulator.ok : 'a -> 'a Accumulator.t`, you could even refactor this as `let return = Fn.compose Deferred.Result.return Accumulator.ok`.